### DISCUSSION TITLE
Trying to communicate via internet

### DISCUSSION THREAD

**Discussion by ygreq** [CESNET/UltraGrid#198] [Q&A]
Hi there!

Newbie questions for sure. I finally got a hold of a 4g router and wanted to test ultragrid over internet. I start with simple things. 

On PC1 I am sending `uv -t  testcard -d gl:d [EXTERNAL_IP_of_PC2]`
On PC2 I am sending `uv -t  testcard -d gl:d [EXTERNAL_IP_of_PC1]`

I opened ports 5004, 5006, 5104, and 5016 on both routers. I used Port Forwarding before so I believe everything is set up right there. 

**The connection does not work. What am I missing?**

I also looked at [NAT traversal](https://github.com/CESNET/UltraGrid/wiki/NAT-traversal), but it's a bit too advanced for me. :/

Any suggestions?

Thank you so much!

**Comment by ygreq** [comment]
BTW, this info appears: `[NAT] Private outbound IPv4 address detected and bound to a non-dynamic port. Consider adding '-N' option for NAT traversal.`

I added -N like this `uv -t testcard -d gl:d [EXTERNAL_IP_of_PC2] -N` but it gives me tons of errors:

```
[NAT] PCP - cannot create video flow!
....
[NAT] PCP initialization failed!
[NAT] NAT PMP - using gateway: 192.168.1.1
[NAT] NAT PMP - readnatpmpresponseorretry returned -100 (TRY AGAIN)
```

**Reply by MartinPulec** [reply]
Hi, hard to say without details. I've pushed a code to checking the timeout from NAT PMP, you can grab the binaries from [there](https://github.com/CESNET/UltraGrid/releases/tag/continuous). It checks timeout setting and exits after 5 second of unsuccessful attempts. It won't fix the problem but may provide more diagnostic info.

Generally saying it is up to router setting if it supports some sort of signalization and is configured to allow it. You can check the router settings (if you have the access) - check for an item labeled (slightly incorrectly) _UPnP_. However if the router doesn't allow any sort of signalization, you'd need to deploy UDP hole punching.

Both computers are behind NAT? If only one, you should definitely try [this](https://github.com/CESNET/UltraGrid/wiki/NAT-traversal#server-mode).

Anyways, if you've set port forwarding manually, the above shouldn't be required. Could you check incoming traffic with eg. tcpdump, jnettop, Wireshark etc.?

**Reply by ygreq** [reply]
Hi Martin! Thank you for all the good advice. What I did not understand from the start is if I need to add **-N** in a normal situation.

**Reply by ygreq** [reply]
I am using the latest version. This is the command `C:\UltraGrid>uv -t testcard -d gl:d [CENSORED_EXTERNAL_IP_of_PC2] -N`
And this is what I get:
```
UltraGrid 1.6+ (tags/continuous rev 698ab753 built Oct 18 2021 08:22:26)

Display device   : gl
Capture device   : testcard
Audio capture    : none
Audio playback   : none
MTU              : 1500 B
Video compression: none
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : none
Video FEC        : none

  0s 000ms 000us INFO   : Found gateway ::ffff:192.168.1.1. Added as possible PCP server.
  0s 000ms 604us INFO   : Found gateway ::ffff:25.255.255.254. Added as possible PCP server.
  0s 000ms 763us INFO   : Added new flow(PCP server: ::ffff:192.168.1.1; Int. addr: [::ffff:192.168.1.CENSORED]:5004; Dest. addr: [::]:0; Key bucket: 13)
  0s 000ms 852us INFO   : Added new flow(PCP server: ::ffff:25.255.255.254; Int. addr: [::ffff:192.168.1.CENSORED]:5004; Dest. addr: [::]:0; Key bucket: 26)
  0s 000ms 938us INFO   : Initialized wait for result of flow: 13, wait timeout 500 ms
  0s 001ms 015us INFO   : Pinging PCP server at address ::ffff:192.168.1.1
  0s 001ms 156us INFO   : Sent PCP MSG (flow bucket:13)
  0s 001ms 212us INFO   : Pinging PCP server at address ::ffff:25.255.255.254
  0s 001ms 313us INFO   : Sent PCP MSG (flow bucket:26)

Flow signaling processing.

Flow signaling processing.
[NAT] PCP - cannot create video flow!
  0s 518ms 660us INFO   : Flow closed(PCP server: ::ffff:192.168.1.1; Int. addr: [::ffff:192.168.1.CENSORED]:5004; Dest. addr: [::]:0; Key bucket: 13)
  0s 518ms 914us INFO   : Sent PCP MSG (flow bucket:13)
  0s 519ms 700us INFO   : Flow closed(PCP server: ::ffff:25.255.255.254; Int. addr: [::ffff:192.168.1.CENSORED]:5004; Dest. addr: [::]:0; Key bucket: 26)
  0s 521ms 035us INFO   : Sent PCP MSG (flow bucket:26)
  0s 521ms 953us INFO   : PCP server ::ffff:192.168.1.1 terminated.
  0s 524ms 536us INFO   : PCP server ::ffff:25.255.255.254 terminated.
[NAT] PCP initialization failed!
[NAT] NAT PMP - using gateway: 192.168.1.1
[NAT] NAT PMP - readnatpmpresponseorretry returned -100 (TRY AGAIN)
    Last message repeated 7 times
[NAT] NAT PMP - readnatpmpresponseorretry returned -7 (FAILED)
[NAT] NAT PMP - readnatpmpresponseorretry() failed : the gateway does not support nat-pmp
[NAT] NAT PMP - errno=0 'No error'
[NAT] NAT PMP initialization failed!
[NAT] Could not initialize any NAT traversal.
Exit
```

**Reply by ygreq** [reply]
Also here you can see the **Port Forwarding** setups on the 2 routers:
TP-Link: https://ibb.co/LChjN0j
Unifi (here I also added port forwarding for a second pc, but changed the ports for it.): https://ibb.co/vJk393w

I also checked UPnP settings on both routers: 
-On TP-Link which is a simple router, I get this: https://ibb.co/G3WT3bw **Should I try to do UDP hole punching in this case?**

-On Unifi, it seems I had the UPnP turned off for security reasons. I turned it on. **And now I believe it is working on that side. 
Unfortunately once UltraGrid started, web browsing stopped. When I stopped UltraGrid, I could use again the browser. I don't know if other services stopped as well.** 

And now I cannot access my Unifi controller. Don't know why. `webRtcId 1 terminated with code: (-2147090409) WebRTC connection interrupted from far side` Anyway I will try to get the unifi controller rolling again and will come with updates. 

I will also check the other things you mentioned

Sorry for double posting. I am trying to keep the thread in order. I deleted the original messages.

**Reply by MartinPulec** [reply]
> Hi Martin! Thank you for all the good advice. What I did not understand from the start is if I need to add **-N** in a normal situation.

Currently this parameter tries to configure the router with PCP/NAT-PMP - no need to do this if port forwarding is set explicitly. The message is just a piece of advice. From your code it looks like the router ignores that control, anyway - either not supporting or not configured. But as I've already written - nothing to worry about if you set the port forwarding manually.

**Reply by MartinPulec** [reply]
> Also here you can see the **Port Forwarding** setups on the 2 routers: TP-Link: https://ibb.co/LChjN0j Unifi (here I also added port forwarding for a second pc, but changed the ports for it.): https://ibb.co/vJk393w
> 
> I also checked UPnP settings on both routers: -On TP-Link which is a simple router, I get this: https://ibb.co/G3WT3bw **Should I try to do UDP hole punching in this case?**

This looks OK, I'd also check the firewall status

> 
> -On Unifi, it seems I had the UPnP turned off for security reasons. I turned it on. **And now I believe it is working on that side. Unfortunately once UltraGrid started, web browsing stopped. When I stopped UltraGrid, I could use again the browser. I don't know if other services stopped as well.**

UltraGrid shouldn't affect anything more than usual. I'd suspect sending an uncompressed stream?

> 
> And now I cannot access my Unifi controller. Don't know why. `webRtcId 1 terminated with code: (-2147090409) WebRTC connection interrupted from far side` Anyway I will try to get the unifi controller rolling again and will come with updates.
> 
> I will also check the other things you mentioned
> 
> Sorry for double posting. I am trying to keep the thread in order. I deleted the original messages.

**Reply by ygreq** [reply]
> This looks OK, I'd also check the firewall status

You mean the firewall on the router or on the pc sending data on that router?

**Reply by ygreq** [reply]
> > -On Unifi, it seems I had the UPnP turned off for security reasons. I turned it on. **And now I believe it is working on that side. Unfortunately once UltraGrid started, web browsing stopped. When I stopped UltraGrid, I could use again the browser. I don't know if other services stopped as well.**
> 
> UltraGrid shouldn't affect anything more than usual. I'd suspect sending an uncompressed stream?

Now it seems to work. I don't know what happened.

**Reply by ygreq** [reply]
Situation right now:

**On laptop on Unifi** I can initiate the UG code both with and without -N. No errors. And I can receive data from the other end. 

**On laptop on TP-Link**, without -N it starts and it sends data to the other end. With -N, it gives me the same error. The data I see on the other end is missing frames, it's corrupt, but still. Probably from the router or the fact that it's a 4G network. Aaand I did not use a codec which I will do from now on. (LE: Yeap! Works perfect with a codec but only one way :D) 

It shows:
```
[testcard] 15 frames in 5.09273 seconds = 2.94537 FPS
[testcard] 14 frames in 5.06578 seconds = 2.76364 FPS
[testcard] 14 frames in 5.06252 seconds = 2.76542 FPS
[testcard] 14 frames in 5.06861 seconds = 2.7621 FPS
[testcard] 14 frames in 5.05042 seconds = 2.77205 FPS
[testcard] 14 frames in 5.06678 seconds = 2.76309 FPS
```

So my guess is that the settings on the TP-Link router are not helping and should be modified. Or the settings on the laptop itself. But I did stop the Windows Firewall and did another test and it still did not work. 

Any suggestions in this situation? 
Thank you!!!

**Reply by ygreq** [reply]
I just tested [Server mode](https://github.com/CESNET/UltraGrid/wiki/NAT-traversal#server-mode) and it looks like it's working. :) Is this one of the normal ways of using UltraGrid or should I struggle to make them work without Server Mode?

**Reply by MartinPulec** [reply]
> **On laptop on TP-Link**, without -N it starts and it sends data to the other end.

Just a clarification - the parameter for _NAT traversal_ (-N) doesn't affect sending in any way. It is mostly useful only for the receiver. (But of course if initialization of NAT traversal fails, it won't send anyway.)

With -N, it gives me the same error. The data I see on the other end is missing frames, it's corrupt, but still. Probably from the router or the fact that it's a 4G network. Aaand I did not use a codec which I will do from now on. (LE: Yeap! Works perfect with a codec but only one way :D)
> 
> It shows:
> 
> ```
> [testcard] 15 frames in 5.09273 seconds = 2.94537 FPS
> [testcard] 14 frames in 5.06578 seconds = 2.76364 FPS
> [testcard] 14 frames in 5.06252 seconds = 2.76542 FPS
> [testcard] 14 frames in 5.06861 seconds = 2.7621 FPS
> [testcard] 14 frames in 5.05042 seconds = 2.77205 FPS
> [testcard] 14 frames in 5.06678 seconds = 2.76309 FPS
> ```
> 
> So my guess is that the settings on the TP-Link router are not helping and should be modified. Or the settings on the laptop itself. But I did stop the Windows Firewall and did another test and it still did not work.
> 
> Any suggestions in this situation? Thank you!!!

What is exactly the outgoing bandwidth utilized by UG here? Hard to say what limits sending - it can be [a flow control](https://en.wikipedia.org/wiki/Ethernet_flow_control). Or just an insufficient performance eg. for uncompressed video.

**Reply by MartinPulec** [reply]
> I just tested [Server mode](https://github.com/CESNET/UltraGrid/wiki/NAT-traversal#server-mode) and it looks like it's working. :) Is this one of the normal ways of using UltraGrid or should I struggle to make them work without Server Mode?

Well, I don't know what is a normal way :) I'd say that ideal is not having a NAT. It is just another workaround to traverse a NAT, I'd say.

**Reply by ygreq** [reply]
I just tested communicating (audio only for now as don't yet have access to 2 fixed internet non-metered connections (one is fixed non-metered and the other is mobile and metered) using [ZeroTier](https://www.zerotier.com/). I set the virtual network IPs and it seems to work without the need for the public IP addresses or setting any NAT traversal options. Do you think there'll be some limitations with this setup? Like I said, I just send minimal data for now and this is why i am asking, but curious if you used ZeroTier or something similar (VPN) in your research. 

Thank you!!

**Reply by MartinPulec** [reply]
I don't know technical details of ZeroTier but if the data plane runs over UDP and the path is direct, I expect it to work correctly - the encapsulation should add only a small overhead and the network should behave in accordance whith the underlying network. I assume that it could behave just similarly to VNC over UDP.
